core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala

```diff
@@ -179,8 +198,13 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
             blocks, SparkEnv.get.serializer, compressionCodec)
           // Store the merged copy in BlockManager so other tasks on this executor don't
           // need to re-fetch it.
-          SparkEnv.get.blockManager.putSingle(
-            broadcastId, obj, StorageLevel.MEMORY_AND_DISK, tellMaster = false)
+          blockManager.putSingle(broadcastId, obj, StorageLevel.MEMORY_AND_DISK, tellMaster = false)
+          Option(TaskContext.get()) match {
+            case Some(taskContext) =>
+              taskContext.addTaskCompletionListener(_ => blockManager.releaseLock(broadcastId))
+            case None =>
+              blockManager.releaseLock(broadcastId)
```

Comment:
in this case we're just putting a block. Why not just always release the block right after the put (if the put is successful)? I don't see any reason to wait until the end of the task.